libsolidity/formal/CHC.cpp

```diff
@@ -40,21 +40,20 @@ CHC::CHC(
 	smt::SMTSolverChoice _enabledSolvers
 ):
 	SMTEncoder(_context),
-#ifdef HAVE_Z3
-	m_interface(
-		_enabledSolvers.z3 ?
-		dynamic_pointer_cast<smt::CHCSolverInterface>(make_shared<smt::Z3CHCInterface>()) :
-		dynamic_pointer_cast<smt::CHCSolverInterface>(make_shared<smt::CHCSmtLib2Interface>(_smtlib2Responses, _smtCallback))
-	),
-#else
-	m_interface(make_shared<smt::CHCSmtLib2Interface>(_smtlib2Responses, _smtCallback)),
-#endif
 	m_outerErrorReporter(_errorReporter),
 	m_enabledSolvers(_enabledSolvers)
 {
-	(void)_smtlib2Responses;
 	(void)_enabledSolvers;
-	(void)_smtCallback;
+	bool usesZ3 = false;
+#ifdef HAVE_Z3
+	if (_enabledSolvers.z3)
+	{
+		usesZ3 = true;
+		m_interface = make_unique<smt::Z3CHCInterface>();
+	}
+#endif
+	if (!usesZ3)
```

Comment:
```suggestion
	if (!m_interface)
```
and remove ``usesZ3`` altogether?

Reply:
done